From ec433d3c8a4c59a762348aabc1cd95030c5be968 Mon Sep 17 00:00:00 2001
From: kargl <kargl@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Thu, 16 Jul 2015 19:22:38 +0000
Subject: [PATCH] 2015-07-16  Steven G. Kargl  <kargl@gcc.gnu.org>

	PR fortran/66864
	* simplify.c (gfc_simplify_floor): Set precision of temporary to
	that of arg.

2015-07-16  Steven G. Kargl  <kargl@gcc.gnu.org>

	PR fortran/66864
	* gfortran.dg/pr66864.f90: New test.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-5-branch@225903 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/fortran/ChangeLog                 |  6 ++++++
 gcc/fortran/simplify.c                |  4 +---
 gcc/testsuite/ChangeLog               |  5 +++++
 gcc/testsuite/gfortran.dg/pr66864.f90 | 16 ++++++++++++++++
 4 files changed, 28 insertions(+), 3 deletions(-)
 create mode 100644 gcc/testsuite/gfortran.dg/pr66864.f90

diff --git a/gcc/fortran/ChangeLog b/gcc/fortran/ChangeLog
index 1133082..7b24278 100644
--- a/gcc/fortran/ChangeLog
+++ b/gcc/fortran/ChangeLog
@@ -1,3 +1,9 @@
+2015-07-16  Steven G. Kargl  <kargl@gcc.gnu.org>
+
+	PR fortran/66864
+	* simplify.c (gfc_simplify_floor): Set precision of temporary to
+	that of arg.
+
 2015-07-16  Steven G. Kargl   <kargl@gcc.gnu.org>
 
 	PR fortran/66545
diff --git a/gcc/fortran/simplify.c b/gcc/fortran/simplify.c
index 20d50d2..e4df72c 100644
--- a/gcc/fortran/simplify.c
+++ b/gcc/fortran/simplify.c
@@ -2352,9 +2352,7 @@ gfc_simplify_floor (gfc_expr *e, gfc_expr *k)
   if (e->expr_type != EXPR_CONSTANT)
     return NULL;
 
-  gfc_set_model_kind (kind);
-
-  mpfr_init (floor);
+  mpfr_init2 (floor, mpfr_get_prec (e->value.real));
   mpfr_floor (floor, e->value.real);
 
   result = gfc_get_constant_expr (BT_INTEGER, kind, &e->where);
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index 0b5de0e..b4d8255 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,8 @@
+2015-07-16  Steven G. Kargl  <kargl@gcc.gnu.org>
+
+	PR fortran/66864
+	* gfortran.dg/pr66864.f90: New test.
+
 2015-07-16  Steven G. Kargl   <kargl@gcc.gnu.org>
 
 	PR fortran/66545
diff --git a/gcc/testsuite/gfortran.dg/pr66864.f90 b/gcc/testsuite/gfortran.dg/pr66864.f90
new file mode 100644
index 0000000..ebea99b
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/pr66864.f90
@@ -0,0 +1,16 @@
+! { dg-do run }
+! PR fortran/66864
+!
+program t
+   implicit none
+   real(8) x
+   x = 2.0d0**26.5d0
+   if (floor(x) /= 94906265) call abort
+   if (floor(2.0d0**26.5d0)/= 94906265) call abort
+   x = 777666555.6d0
+   if (floor(x) /= 777666555) call abort
+   if (floor(777666555.6d0) /= 777666555) call abort
+   x = 2000111222.6d0
+   if (floor(x) /= 2000111222) call abort
+   if (floor(2000111222.6d0) /= 2000111222) call abort
+end program t
-- 
1.9.4

